### ISSUE TITLE
False negative lint for `unused_imports`

### ISSUE THREAD

**Issue by bvanjoi** [rust-lang/rust#126122]
### Code

```Rust
mod a {
  pub(crate) use crate::s;
}
mod b {
  pub mod s {}
}
use self::b::*;
use self::a::s;

fn main() {}
```

### Current output

```Shell
nothing
```

### Desired output

```Shell
Report `use self::a::s` as `unused_import`
```

### Rationale and extra context

If you make a slight modification:

```diff
mod a {
  pub(crate) use crate::s;
}
mod b {
  pub mod s {}
}
use self::b::*;
- use self::a::s;
+ use self::a::s as other_s;

fn main() {}
```

It will report ﻿`use self::a::s as other_s` as an unused import.

### Rust Version

```Shell
rustc 1.80.0-nightly (72fdf913c 2024-06-05)
binary: rustc
commit-hash: 72fdf913c53dd0e75313ba83e4aa80df3f6e2871
commit-date: 2024-06-05
host: aarch64-apple-darwin
release: 1.80.0-nightly
LLVM version: 18.1.6
```

**Comment by kristof-mattei** [comment]
Another false negative:

```rust
#![deny(unused)]

use tracing::warn;

fn main() {
    #[warn(clippy::suspicious_else_formatting)]
    {}
}
```

I would expect the above to hit the deny, because we're not using `tracing::warn`. 

Commenting `#[warn(clippy::suspicious_else_formatting)]` works:

```rust
    Checking playground v0.0.1 (/playground)
error: unused import: `tracing::warn`
 --> src/main.rs:3:5
  |
3 | use tracing::warn;
  |     ^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(unused)]
  |         ^^^^^^
  = note: `#[deny(unused_imports)]` implied by `#[deny(unused)]`

error: could not compile `playground` (bin "playground") due to 1 previous error
```

Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=dfb7aff960b78e80e42ae48e8649e81d
